zapgrpc/zapgrpc.go

```diff
@@ -36,65 +59,148 @@ func (f optionFunc) apply(log *Logger) {
+	}
 }
 
-// Fatalf implements grpclog.Logger.
-func (l *Logger) Fatalf(format string, args ...interface{}) {
-	l.fatalf(l.log, format, args...)
+// Printf implements grpclog.Logger.
+// Deprecated: use Infof().
+func (l *Logger) Printf(format string, args ...interface{}) {
+	if l.printToDebug {
+		l.delegate.Debugf(format, args...)
+	} else {
+		l.delegate.Infof(format, args...)
+	}
+}
+
+// Println implements grpclog.Logger.
+// Deprecated: use Info().
+func (l *Logger) Println(args ...interface{}) {
+	l.Print(addSpaces(args)...)
+}
+
+// Info implements grpclog.LoggerV2.
+func (l *Logger) Info(args ...interface{}) {
+	l.delegate.Info(args...)
+}
+
+// Infoln implements grpclog.LoggerV2.
+func (l *Logger) Infoln(args ...interface{}) {
+	l.delegate.Info(addSpaces(args)...)
+}
+
+// Infof implements grpclog.LoggerV2.
+func (l *Logger) Infof(format string, args ...interface{}) {
+	l.delegate.Infof(format, args...)
+}
+
+// Warning implements grpclog.LoggerV2.
+func (l *Logger) Warning(args ...interface{}) {
+	l.delegate.Warn(args...)
 }
 
-// Fatalln implements grpclog.Logger.
+// Warningln implements grpclog.LoggerV2.
+func (l *Logger) Warningln(args ...interface{}) {
+	l.delegate.Warn(addSpaces(args)...)
+}
+
+// Warningf implements grpclog.LoggerV2.
+func (l *Logger) Warningf(format string, args ...interface{}) {
+	l.delegate.Warnf(format, args...)
+}
+
+// Error implements grpclog.LoggerV2.
+func (l *Logger) Error(args ...interface{}) {
+	l.delegate.Error(args...)
+}
+
+// Errorln implements grpclog.LoggerV2.
+func (l *Logger) Errorln(args ...interface{}) {
+	l.delegate.Error(addSpaces(args)...)
+}
+
+// Errorf implements grpclog.LoggerV2.
+func (l *Logger) Errorf(format string, args ...interface{}) {
+	l.delegate.Errorf(format, args...)
+}
+
+// Fatal implements grpclog.LoggerV2.
+func (l *Logger) Fatal(args ...interface{}) {
+	if l.fatalToWarn {
+		l.delegate.Warn(args...)
+	} else {
+		l.delegate.Fatal(args...)
+	}
+}
+
+// Fatalln implements grpclog.LoggerV2.
 func (l *Logger) Fatalln(args ...interface{}) {
-	l.fatal(l.log, args...)
+	l.Fatal(addSpaces(args)...)
 }
 
-// Print implements grpclog.Logger.
-func (l *Logger) Print(args ...interface{}) {
-	l.print(l.log, args...)
+// Fatalf implements grpclog.LoggerV2.
+func (l *Logger) Fatalf(format string, args ...interface{}) {
+	if l.fatalToWarn {
+		l.delegate.Warnf(format, args...)
+	} else {
+		l.delegate.Fatalf(format, args...)
+	}
 }
 
-// Printf implements grpclog.Logger.
-func (l *Logger) Printf(format string, args ...interface{}) {
-	l.printf(l.log, format, args...)
+// V implements grpclog.LoggerV2.
+func (l *Logger) V(level int) bool {
+	return l.delegate.Desugar().Core().Enabled(grpc2zapLvl[level])
 }
 
-// Println implements grpclog.Logger.
-func (l *Logger) Println(args ...interface{}) {
-	l.print(l.log, args...)
+// addSpaces always adds spaces between arguments like https://golang.org/pkg/fmt/#Println
+func addSpaces(args []interface{}) []interface{} {
```

Comment:
instead of this, we could use `Sprintln(args)` and trim the trailing newline and use the string instead

Reply:
We could, but I think this would lead to more memory allocations. This code only allocates once, only when needed and the perfect slice size. `Sprintln()` would turn all arguments into strings, which is arbitrarily expensive in terms of both CPU and memory. I think it'd better to keep the current code.

Reply:
We have to call `Sprint*` at some point to convert the arguments into a string for the logger API. With the current implementation, doing `Sprintln` here would be an additional allocation, but we can actually avoid a second `Sprint` in the case we receive a single string argument,
https://github.com/uber-go/zap/pull/913

With the above change merged in, `Sprintln` will actually be more efficient, as we convert the arguments to a string once (1 alloc), vs creating a slice here with the strings, and then converting them using `Sprint` later (2 allocs).

Reply:
The above is true, but consider a case where arguments are expensive to turn into strings and the logger is disabled. This is a common scenario - verbose debug logging that outputs bulky objects with all the details. This case would perform much worse if we use `Sprint` here.

As a user, I would expect the logger to evaluate (i.e. turn into strings) arguments if and only if it will then actually log. As a user, I would not expect to need to guard my log statements with `if is_log_level_enabled` checks but just rely on the above.

Reply:
Agree that we should avoid expensive operations if the logger is disabled (without pushing any checks to the caller). However, this implementation still allocates for the slice even if the log is disabled. An alternate option is to add level checks into the gRPC logger explicitly, so it skips doing any work if the level is disabled.  If the level is enabled, then we can use `Sprintln` to serialize the arguments. Would that work?

Reply:
```go
// Fatalln implements grpclog.LoggerV2.
func (l *Logger) Fatalln(args ...interface{}) {
	if l.delegate.Desugar().Core().Enabled(zapcore.FatalLevel) {
		l.Fatal(fmt.Sprintln(args...))
	}
}
```
This would allocate too in the `Desugar()` call 🤷 I've made the change, PTAL.

Reply:
I've extracted the `LevelEnabler` into a field to avoid this allocation.